### PR TITLE
257 verify loop fusion

### PR DESCRIPTION
**Pull request by hiker**

### CODE REVIEW

**Comment by hiker** [comment]
I've found an additional problem with the loop fuse: the loops can be specified in the wrong order, and this will be accepted by validate. But during fuse the loop bodies will be ending up in the wrong order!

**Comment by hiker** [comment]
Ready for a first review. I have moved the dependency analysis required in the validation function into the DependencyTool , anticipating the need to use the existing code for DA. Unfortunately (and as discussed), I don't have time atm to implement the full DA for fusion, I only verify that all accesses in the fused dimension are consistent (i.e. no stencil read access for variables that are also written) - this can result in some cases to reject fusion when it is still valid. But we have a 'force' option so we can overwrite this (and much better to reject a bit too keen compared with the current state where we incorrectly fuse).

Note that in order to support renaming of loop variables I had to add support to SymPy to indicate that some variables have the same value.

I also moved most of the failing loop fusion tests into the dependency tools test (to reach 100% coverage of the latter).

I've triggered the CI as well.

File: src/psyclone/core/symbolic_maths.py [hunk 1]
```diff
@@ -82,13 +82,18 @@ def get():
 
     # -------------------------------------------------------------------------
     @staticmethod
-    def equal(exp1, exp2):
+    def equal(exp1, exp2, assume=None):
         '''Test if the two PSyIR expressions are symbolically equivalent.
+        The optional assume dictionary can contain information which variables
+        are known to be the same. For example, if `assume={'i': 'j'}`, then
```
Comment (arporter) [inline src/psyclone/core/symbolic_maths.py:88]: I don't think 'assume' is quite the right name here. Perhaps "identical_variables" or something?

File: src/psyclone/core/symbolic_maths.py [hunk 2]
```diff
@@ -164,7 +169,7 @@ def never_equal(exp1, exp2):
 
     # -------------------------------------------------------------------------
     @staticmethod
-    def _subtract(exp1, exp2):
+    def _subtract(exp1, exp2, assume=None):
```
Comment (arporter) [inline src/psyclone/core/symbolic_maths.py:172]: Again, please rename this argument.

File: src/psyclone/core/symbolic_maths.py [hunk 3]
```diff
@@ -82,13 +82,18 @@ def get():
...
+        The optional assume dictionary can contain information which variables
+        are known to be the same. For example, if `assume={'i': 'j'}`, then
+        'i+1' and 'j+1' will be considered equal.
 
         :param exp1: the first expression to be compared.
         :type exp1: :py:class:`psyclone.psyir.nodes.Node`
         :param exp2: the second expression to be compared.
         :type exp2: :py:class:`psyclone.psyir.nodes.Node`
+        :param assume: which variable names are known to be identical
+        :type assume: dict[str, str]
```
Comment (arporter) [inline src/psyclone/core/symbolic_maths.py:96]: `Optional[dict[...]]`

File: src/psyclone/core/symbolic_maths.py [hunk 4]
```diff
@@ -175,6 +180,8 @@ def _subtract(exp1, exp2):
         :type exp1: Optional[:py:class:`psyclone.psyir.nodes.Node`]
         :param exp2: the second expression to be compared.
         :type exp2: Optional[:py:class:`psyclone.psyir.nodes.Node`]
+        :param assume: which variable names are known to be identical
+        :type assume: dict[str, str]
```
Comment (arporter) [inline src/psyclone/core/symbolic_maths.py:184]: `Optional[...]`

File: src/psyclone/psyir/backend/sympy_writer.py [hunk 5]
```diff
@@ -246,7 +246,7 @@ def _create_sympy_array_function(self, name, sig=None, num_dims=None):
         return new_func
 
     # -------------------------------------------------------------------------
-    def _create_type_map(self, list_of_expressions):
+    def _create_type_map(self, list_of_expressions, assume=None):
```
Comment (arporter) [inline src/psyclone/psyir/backend/sympy_writer.py:249]: Please rename.

File: src/psyclone/psyir/backend/sympy_writer.py [hunk 6]
```diff
@@ -269,6 +269,13 @@ def _create_type_map(self, list_of_expressions):
         ``lambda_1``). The SymPyWriter (e.g. in ``reference_node``) will
         use the renamed value when creating the string representation.
 
+        The optional assume dictionary can contain information which variables
```
Comment (arporter) [inline src/psyclone/psyir/backend/sympy_writer.py:272]: "...information about any variables which are known..."

File: src/psyclone/psyir/backend/sympy_writer.py [hunk 7]
```diff
@@ -269,6 +269,13 @@ def _create_type_map(self, list_of_expressions):
...
+        are known to be the same. For example, if `assume={'i': 'j'}`, then
+        'i+1' and 'j+1' will be considered equal.
+
+        :param assume: which variable names are known to be identical
+        :type assume: dict[str, str]
```
Comment (arporter) [inline src/psyclone/psyir/backend/sympy_writer.py:277]: `Optional[]`

File: src/psyclone/psyir/backend/sympy_writer.py [hunk 8]
```diff
@@ -310,7 +317,8 @@ def _create_type_map(self, list_of_expressions):
                 unique_sym = self._symbol_table.new_symbol(name, tag=name)
                 # Test if an array or an array expression is used:
                 if not ref.is_array:
-                    self._sympy_type_map[unique_sym.name] = Symbol(name)
+                    if unique_sym.name not in self._sympy_type_map:
```
Comment (arporter) [inline src/psyclone/psyir/backend/sympy_writer.py:320]: Why is this change needed? Is it a bug fix or something else?

File: src/psyclone/psyir/backend/sympy_writer.py [hunk 9]
```diff
@@ -358,11 +376,17 @@ def type_map(self):
         return self._sympy_type_map
 
     # -------------------------------------------------------------------------
-    def _to_str(self, list_of_expressions):
+    def _to_str(self, list_of_expressions, assume=None):
```
Comment (arporter) [inline src/psyclone/psyir/backend/sympy_writer.py:379]: Please rename.

File: src/psyclone/psyir/backend/sympy_writer.py [hunk 10]
```diff
@@ -358,11 +376,17 @@ def type_map(self):
...
         '''Converts PSyIR expressions to strings. It will replace Fortran-
         specific expressions with code that can be parsed by SymPy. The
         argument can either be a single element (in which case a single string
         is returned) or a list/tuple, in which case a list is returned.
+        The optional assume dictionary can contain information which variables
```
Comment (arporter) [inline src/psyclone/psyir/backend/sympy_writer.py:384]: "...information about which variables..."

File: src/psyclone/psyir/backend/sympy_writer.py [hunk 11]
```diff
@@ -358,11 +376,17 @@ def type_map(self):
...
+        are known to be the same. For example, if `assume={'i': 'j'}`, then
+        'i+1' and 'j+1' will be considered equal.
+
+        :param assume: which variable names are known to be identical
+        :type assume: dict[str, str]
```
Comment (arporter) [inline src/psyclone/psyir/backend/sympy_writer.py:389]: `Optional[...]`

File: src/psyclone/psyir/backend/sympy_writer.py [hunk 12]
```diff
@@ -392,19 +416,24 @@ def _to_str(self, list_of_expressions):
...
-    def __call__(self, list_of_expressions):
+    def __call__(self, list_of_expressions, assume=None):
         '''
         This function takes a list of PSyIR expressions, and converts
         them all into Sympy expressions using the SymPy parser.
         It takes care of all Fortran specific conversion required (e.g.
         constants with kind specification, ...), including the renaming of
         member accesses, as described in
         https://psyclone-dev.readthedocs.io/en/latest/sympy.html#sympy
+        The optional assume dictionary can contain information which variables
```
Comment (arporter) [inline src/psyclone/psyir/backend/sympy_writer.py:427]: "about which"

File: src/psyclone/psyir/backend/sympy_writer.py [hunk 13]
```diff
@@ -392,19 +416,24 @@ def _to_str(self, list_of_expressions):
         return expression_str_list
 
     # -------------------------------------------------------------------------
-    def __call__(self, list_of_expressions):
+    def __call__(self, list_of_expressions, assume=None):
```
Comment (arporter) [inline src/psyclone/psyir/backend/sympy_writer.py:419]: Please re-name.

File: src/psyclone/psyir/backend/sympy_writer.py [hunk 14]
```diff
@@ -392,19 +416,24 @@ def _to_str(self, list_of_expressions):
...
+        The optional assume dictionary can contain information which variables
+        are known to be the same. For example, if `assume={'i': 'j'}`, then
+        'i+1' and 'j+1' will be considered equal.
 
         :param list_of_expressions: the list of expressions which are to be
             converted into SymPy-parsable strings.
         :type list_of_expressions: list of
             :py:class:`psyclone.psyir.nodes.Node`
+        :param assume: which variable names are known to be identical
+        :type assume: dict[str, str]
```
Comment (arporter) [inline src/psyclone/psyir/backend/sympy_writer.py:436]: `Optional[...]`

File: src/psyclone/psyir/backend/sympy_writer.py [hunk 15]
```diff
@@ -418,7 +447,7 @@ def __call__(self, list_of_expressions):
         is_list = isinstance(list_of_expressions, (tuple, list))
```
Comment (arporter) [inline src/psyclone/psyir/backend/sympy_writer.py:447]: Since this is the entry point at which the caller first supplies `assumes`, we should do some type-checking here.

File: src/psyclone/psyir/tools/dependency_tools.py [hunk 16]
```diff
@@ -806,3 +811,208 @@ def can_loop_be_parallelised(self, loop,
...
 
         return result
+
+# -------------------------------------------------------------------------
+    def can_loops_be_fused(self, loop1, loop2):
+        '''Function that verifies if two loops can be fused.
+
+        :param loop1: the first loop.
+        :type loop1: :py:class:`psyclone.psyir.nodes.Loop`
+        :param loop1: the first loop.
```
Comment (arporter) [inline src/psyclone/psyir/tools/dependency_tools.py:821]: "loop2" and "the second loop" :-) Same for the `type` below.

File: src/psyclone/psyir/tools/dependency_tools.py [hunk 17]
```diff
@@ -806,3 +811,208 @@ def can_loop_be_parallelised(self, loop,
...
+        :return: whether the loops can be fused or not.
+        :rtype: bool
+
+        '''
+        # pylint: disable=too-many-locals
+        # This should only be called from loop_fuse_trans, which
+        # has done tests for loop boundaries (depending on domain)
+
+        self._clear_messages()
+        assert isinstance(loop1, Loop)
```
Comment (arporter) [inline src/psyclone/psyir/tools/dependency_tools.py:833]: Either we assume that the arguments are correct (which I'm prepared to accept) or we should check properly and raise an appropriate error.

File: src/psyclone/psyir/transformations/loop_fuse_trans.py [hunk 18]
```diff
@@ -118,6 +117,13 @@ def validate(self, node1, node2, options=None):
             raise TransformationError(
                 f"Error in {self.name} transformation. Nodes are not siblings "
                 f"who are next to each other.")
+
+        # Check node1 comes before node2:
+        if node2.position-node1.position != 1:
+            raise TransformationError(
+                f"Error in {self.name} transformation. The second loop comes "
```
Comment (arporter) [inline src/psyclone/psyir/transformations/loop_fuse_trans.py:124]: The check is more general than the error message says. Perhaps: "the second loop does not immediately follow the first loop"?

File: src/psyclone/psyir/tools/dependency_tools.py [hunk 19]
```diff
@@ -806,3 +811,208 @@ def can_loop_be_parallelised(self, loop,
...
+                          f"written in one loop, but only read in the "
+                          f"other loop.", DTCode.ERROR_SCALAR_WRITTEN_AND_READ)
+        return False
+
+    # -------------------------------------------------------------------------
+    def _fuse_validate_written_array(self, var_info1, var_info2,
+                                     loop_variable1, loop_variable2):
+        '''Validates if the accesses to an array, which is at least written
+        once, allows loop fusion. The access pattern to this array is
+        specified in the two parameters `var_info1` and `var_info2`. Ff
```
Comment (arporter) [inline src/psyclone/psyir/tools/dependency_tools.py:938]: s/Ff/If/

File: src/psyclone/psyir/tools/dependency_tools.py [hunk 20]
```diff
@@ -806,3 +811,208 @@ def can_loop_be_parallelised(self, loop,
...
+                    break
+            else:
+                error = (f"Variable '{var_info1.signature[0]}' does not "
+                         f"depend on loop variable '{loop_var_name1}', but is "
+                         f"read and written")
+                self._add_message(error, DTCode.ERROR_READ_WRITE_NO_LOOP_VAR,
+                                  [var_info1.signature[0]])
+                return False
+
+            # If the loop variable contains more than one index, it is
```
Comment (arporter) [inline src/psyclone/psyir/tools/dependency_tools.py:988]: I don't understand what this means - how can a loop variable contain an index?

File: src/psyclone/psyir/tools/dependency_tools.py [hunk 21]
```diff
@@ -806,3 +811,208 @@ def can_loop_be_parallelised(self, loop,
...
+
+        '''
+        # pylint: disable=too-many-locals
+        # This should only be called from loop_fuse_trans, which
+        # has done tests for loop boundaries (depending on domain)
+
+        self._clear_messages()
+        assert isinstance(loop1, Loop)
+        assert isinstance(loop2, Loop)
+        vars1 = VariablesAccessInfo(loop1)
```
Comment (arporter) [inline src/psyclone/psyir/tools/dependency_tools.py:835]: I've forgotten how this works for LFRic. Does a kernel with a stencil  access (specified in its metadata) end up with a 'fake' array access created for the appropriate field argument?

File: src/psyclone/tests/psyir/backend/sympy_writer_test.py [hunk 22]
```diff
@@ -251,7 +251,7 @@ def test_sym_writer_rename_members(fortran_reader, expressions):
 
 
 @pytest.mark.parametrize("expr, sym_map", [("a%x", {"a": Symbol("a"),
-                                                    "a_x": Symbol("a_x")}),
+                                                    "a_x": Symbol("a%x")}),
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/sympy_writer_test.py:254]: Bug fix?

File: src/psyclone/tests/psyir/backend/sympy_writer_test.py [hunk 23]
```diff
@@ -554,3 +554,36 @@ def test_sym_writer_reserved_names(fortran_reader, expression):
...
+    '''
+    # A dummy program to easily create the PSyIR for the
+    # expressions we need. We just take the RHS of the assignments
+    source = f'''program test_prog
+                use some_mod
+                integer :: a, b, c
+                x = {expressions[0]}
+                end program test_prog '''
+    psyir = fortran_reader.psyir_from_source(source)
+    # psyir is a FileContainer, its first child the program, and its
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/sympy_writer_test.py:583]: More robust to do `psyir.walk(Assignment)[0].rhs`

File: src/psyclone/tests/psyir/backend/sympy_writer_test.py [hunk 24]
```diff
@@ -554,3 +554,36 @@ def test_sym_writer_reserved_names(fortran_reader, expression):
...
+    # first child the assignment, of which we take the right hand side
+    expr = psyir.children[0].children[0].rhs
+
+    sympy_writer = SymPyWriter()
+    assume = {'a': 'b'}
+    assert str(sympy_writer(expr, assume=assume)) == expressions[1]
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/sympy_writer_test.py:589]: Nice.

File: src/psyclone/tests/psyir/tools/dependency_tools_test.py [hunk 25]
```diff
@@ -41,10 +41,17 @@
...
 from psyclone.core import Signature, VariablesAccessInfo
 from psyclone.errors import InternalError
+from psyclone.psyir.nodes import Loop
 from psyclone.psyir.tools import DependencyTools, DTCode
 from psyclone.tests.utilities import get_invoke
 
 
+@pytest.fixture(scope="function", autouse=True)
+def clear_config_instance():
+    '''The tests in this file all assume that the Nemo API is used.'''
```
Comment (arporter) [inline src/psyclone/tests/psyir/tools/dependency_tools_test.py:51]: "...all assume that no DSL API is used."

File: src/psyclone/tests/psyir/tools/dependency_tools_test.py [hunk 26]
```diff
@@ -729,3 +736,305 @@ def test_gocean_parallel():
             "'u_fld(i,j - 1)' in '< kern call: stencil_not_parallel_code >' "
             "are dependent and cannot be parallelised. Variable: 'u_fld'."
             in str(dep_tools.get_all_messages()[0]))
+
+
+def test_fuse_different_variables_with_access(fortran_reader):
+    '''Test that fusing loops with different variables is disallowed when
+    either loop uses the other loops variable for any reason.'''
```
Comment (arporter) [inline src/psyclone/tests/psyir/tools/dependency_tools_test.py:743]: "loop's"

File: src/psyclone/tests/psyir/tools/dependency_tools_test.py [hunk 27]
```diff
@@ -729,3 +736,305 @@ def test_gocean_parallel():
...
+    dep_tools = DependencyTools()
+
+    assert not dep_tools.can_loops_be_fused(loops[1], loops[2])
+    assert len(dep_tools.get_all_messages()) == 1
+    msg = dep_tools.get_all_messages()[0]
+    assert ("Second loop contains accesses to the first loop's variable: ji"
+            in str(msg))
+    assert msg.var_names[0] == "ji"
+
+    assert not dep_tools.can_loops_be_fused(loops[2], loops[1])
```
Comment (arporter) [inline src/psyclone/tests/psyir/tools/dependency_tools_test.py:768]: Shouldn't this be rejected because the loops are supplied in the wrong order?

File: src/psyclone/tests/psyir/tools/dependency_tools_test.py [hunk 28]
```diff
@@ -729,3 +736,305 @@ def test_gocean_parallel():
...
+    msg = dep_tools.get_all_messages()[0]
+    assert ("First loop contains accesses to the second loop's variable: ji"
+            in str(msg))
+    assert msg.var_names[0] == "ji"
+
+
+# ----------------------------------------------------------------------------
+def test_fuse_independent_array_stencil(fortran_reader):
+    '''Test that using arrays which are not dependent on the loop variable
```
Comment (arporter) [inline src/psyclone/tests/psyir/tools/dependency_tools_test.py:777]: This docstring and the comment below don't seem to match what is tested. Does the test title need changing too?

File: src/psyclone/tests/psyir/tools/dependency_tools_test.py [hunk 29]
```diff
@@ -729,3 +736,305 @@ def test_gocean_parallel():
...
+
+# ----------------------------------------------------------------------------
+def test_fuse_independent_array(fortran_reader):
+    '''Test that using arrays which are not dependent on the loop variable
+    are handled correctly. Example:
+    do j  ... a(1) = b(j) * c(j)
+    do j ...  d(j) = a(1)
+    '''
+
+    # The first example can be merged, since 's' does not
```
Comment (arporter) [inline src/psyclone/tests/psyir/tools/dependency_tools_test.py:819]: Again, comment (and docstring and title?) don't reflect what is being tested.

File: src/psyclone/tests/psyir/tools/dependency_tools_test.py [hunk 30]
```diff
@@ -729,3 +736,305 @@ def test_gocean_parallel():
...
+              end subroutine sub'''
+    psyir = fortran_reader.psyir_from_source(code)
+    loop1 = psyir.children[0].children[0]
+    loop2 = psyir.children[0].children[1]
+    assert dep_tools.can_loops_be_fused(loop1, loop2)
+
+
+# ----------------------------------------------------------------------------
+def test_fuse_dimension_change(fortran_reader):
+    '''Test that inconsistent use of dimemsions are detected, e.g.:
```
Comment (arporter) [inline src/psyclone/tests/psyir/tools/dependency_tools_test.py:952]: "dimensions"

File: src/psyclone/tests/psyir/transformations/loop_fusion_test.py [hunk 31]
```diff
@@ -146,6 +146,36 @@ def fuse_loops(fortran_code, fortran_reader, fortran_writer):
     return fortran_writer(psyir), psyir
 
 
+# ----------------------------------------------------------------------------
+def test_fuse_dependency_tools_called(fortran_reader):
+    '''Make sure that DepencendyTools.can_loops_be_fused is indeed called
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/loop_fusion_test.py:151]: "DependencyTools"

File: src/psyclone/tests/psyir/transformations/loop_fusion_test.py [hunk 32]
```diff
@@ -146,6 +146,36 @@ def fuse_loops(fortran_code, fortran_reader, fortran_writer):
...
+              do jj=1, n
+                 t(jj) = t(jj) - a
+              enddo
+              end subroutine sub'''
+    psyir = fortran_reader.psyir_from_source(code)
+    # Get the PSyIR of the original loops
+    loop1 = psyir.children[0].children[0]
+    loop2 = psyir.children[0].children[1]
+
+    fuse = LoopFuseTrans()
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/loop_fusion_test.py:172]: You could monkeypatch `DependencyTools.can_loops_be_fused` for this test instead?

File: src/psyclone/tests/psyir/transformations/loop_fusion_test.py [hunk 33]
```diff
@@ -332,10 +362,12 @@ def test_fuse_correct_bounds(tmpdir, fortran_reader, fortran_writer):
 
 # ----------------------------------------------------------------------------
 def test_fuse_dimension_change(tmpdir, fortran_reader, fortran_writer):
-    '''Test that inconsistent use of dimemsions are detected, e.g.:
+    '''Test that inconsistent use of dimensions are detected, e.g.:
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/loop_fusion_test.py:365]: Strictly, this test now only checks in the case of consistent use of dimensions because, as you say, the rest has been moved.

File: src/psyclone/tests/psyir/transformations/loop_fusion_test.py [hunk 34]
```diff
@@ -368,114 +400,13 @@ def test_fuse_dimension_change(tmpdir, fortran_reader, fortran_writer):
...
-    assert ("Variable 's' does not depend on loop variable 'jj', but is "
-            "read and written" in str(err.value))
-
 
 # ----------------------------------------------------------------------------
 def test_fuse_scalars(tmpdir, fortran_reader, fortran_writer):
     '''Test that using scalars work as expected in all combinations of
-    being read/written in both loops.
+    being read/written in both loops. Note that the dependency tools check
+    more cases, this test is only her to check compilation of the fused
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/loop_fusion_test.py:408]: s/her/here/

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good Joerg, thanks for doing this. Functionally I think it's all find - I just have quibbles about naming and docstrings :-) I'll check the reference guide builds next time.

**Comment by hiker** [comment]
Ready for next review. I've triggered the CI, since I'd guess there is a good chance for this change affecting some transformations.

File: src/psyclone/psyir/tools/dependency_tools.py (hunk 21 above)
Comment (arporter) [reply src/psyclone/psyir/tools/dependency_tools.py:833]: I'm sure I saw some emails with some responses to this review but GH isn't showing me anything now. Please could you comment on this comment?

File: src/psyclone/tests/psyir/tools/dependency_tools_test.py [hunk 35]
```diff
@@ -773,15 +776,11 @@ def test_fuse_different_variables_with_access(fortran_reader):
 
 
 # ----------------------------------------------------------------------------
-def test_fuse_independent_array_stencil(fortran_reader):
-    '''Test that using arrays which are not dependent on the loop variable
-    are handled correctly. Example:
-    do j  ... a(1) = b(j) * c(j)
-    do j ...  d(j) = a(1)
+def test_fuse_inconsistent_array_stencil(fortran_reader):
```
Comment (arporter) [inline src/psyclone/tests/psyir/tools/dependency_tools_test.py:779]: s/_stencil/_indexing/

**Review by arporter — CHANGES_REQUESTED** [review]
Almost there now. Just a bit more tidying to do.
Reference guide builds fine.
Integration tests were fine.

File: src/psyclone/psyir/backend/sympy_writer.py [hunk 36]
```diff
@@ -442,12 +449,25 @@ def __call__(self, list_of_expressions, assume=None):
                       List[:py:class:`sympy.core.basic.Basic`]]
 
         :raises VisitorError: if an invalid SymPy expression is found.
+        :raises InternalError: if the identical_variables parameter is not
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/psyir/backend/sympy_writer.py:452]: Done

File: src/psyclone/tests/psyir/tools/dependency_tools_test.py (hunk 35 above)
Comment (hiker) [reply src/psyclone/tests/psyir/tools/dependency_tools_test.py:779]: Done.

File: src/psyclone/psyir/tools/dependency_tools.py (hunk 21 above)
Comment (hiker) [reply src/psyclone/psyir/tools/dependency_tools.py:833]: @arporter Dang, I was about to say I don't know what you mean, when I realised I have a 'review' pending (answering as 'review' means less emails). So, I'll click on finish review now, and hope that wonderful things will happen.

**Comment by hiker** [comment]
Looks like I forgot to update this PR. I had triggered the CI a few days ago, ready for next review.

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Ref. Guide still builds fine.
I've brought the branch up to master (minor changes) so will do a final check on the integration tests before merging.
